### PR TITLE
Preliminary Munit port

### LINKED ISSUE #3534: Consider migrating to Munit
ScalaTest is making it quite hard to keep up with the dotty milestones, we really don't need any of the functionality that ScalaTest provides, so migrating to Munit should be fairly straightforward, even if it's a lot of work.

### PR DESCRIPTION
**Pull request by LukaJCB**
Should resolve #3534

### CODE REVIEW

File: tests/src/test/scala-2.13+/cats/tests/ArraySeqSuite.scala [hunk 1]
```diff
@@ -60,13 +60,13 @@ class ArraySeqSuite extends CatsSuite {
   }
 
   test("show") {
-    ArraySeq(1, 2, 3).show should ===(s"ArraySeq(1, 2, 3)")
-    ArraySeq.empty[Int].show should ===(s"ArraySeq()")
+assert(    ArraySeq(1, 2, 3).show === (s"ArraySeq(1, 2, 3)"))
+assert(    ArraySeq.empty[Int].show === (s"ArraySeq()"))
```
Comment (djspiewak) [inline tests/src/test/scala-2.13+/cats/tests/ArraySeqSuite.scala:64]: o.O

Comment (LukaJCB) [reply tests/src/test/scala-2.13+/cats/tests/ArraySeqSuite.scala:64]: Ah damnit, need to run scalafmt for 2.13

Comment (gabro) [reply tests/src/test/scala-2.13+/cats/tests/ArraySeqSuite.scala]: by the way, you could think of introducing a CatsSuite trait which defines an assertEq method using Eq. This pattern is how we use MUnit in Metals, where we have a few custom assertion methods built on top of the stock ones

File: tests/src/test/scala/cats/tests/NonEmptySetSuite.scala [hunk 2]
```diff
@@ -216,7 +216,7 @@ class NonEmptySetSuite extends CatsSuite {
 
   test("NonEmptySet#zipWithIndex is consistent with Set#zipWithIndex") {
     forAll { (nes: NonEmptySet[Int]) =>
-      assert(nes.zipWithIndex.toSortedSet === (nes.toSortedSet.zipWithIndex))
+      assert(nes.toSortedSet.zipWithIndex === nes.zipWithIndex.toSortedSet)
```
Comment (larsrh) [inline tests/src/test/scala/cats/tests/NonEmptySetSuite.scala:219]: Why is the equality flipped?

File: build.sbt [hunk 3]
```diff
@@ -126,6 +127,7 @@ lazy val commonJsSettings = Seq(
   jsEnv := new org.scalajs.jsenv.nodejs.NodeJSEnv(),
   // batch mode decreases the amount of memory needed to compile Scala.js code
   scalaJSLinkerConfig := scalaJSLinkerConfig.value.withBatchMode(isTravisBuild.value),
+  scalaJSLinkerConfig ~= (_.withModuleKind(ModuleKind.CommonJSModule)),
```
Comment (larsrh) [inline build.sbt:130]: Why is this change necessary?

File: core/src/main/scala/cats/Traverse.scala [hunk 4]
```diff
@@ -133,7 +133,6 @@ import scala.annotation.implicitNotFound
 }
 
 object Traverse {
-  implicit def catsTraverseForEither[A]: Traverse[Either[A, *]] = cats.instances.either.catsStdInstancesForEither[A]
```
Comment (larsrh) [inline core/src/main/scala/cats/Traverse.scala:136]: Why is this line gone?

File: build.sbt (hunk 3 above)
Comment (LukaJCB) [reply build.sbt:130]: I honestly don't know, but the error message I got when running the tests without it pointed me to this. :smile:

File: core/src/main/scala/cats/Traverse.scala (hunk 4 above)
Comment (LukaJCB) [reply core/src/main/scala/cats/Traverse.scala:136]: I think this line is a mistake (it should be in `UnorderedFoldable`) and it actually caused an issue when compiling with dotty 0.25. I must've checked this in by mistake, I'll include it in a followup PR for dotty.

Comment (larsrh) [reply core/src/main/scala/cats/Traverse.scala:136]: Alright, but isn't this a binary-breaking change?

File: tests/src/test/scala/cats/tests/NonEmptySetSuite.scala (hunk 2 above)
Comment (LukaJCB) [reply tests/src/test/scala/cats/tests/NonEmptySetSuite.scala:219]: I'ts because `===` in ScalaTest doesn't require type equality, whereas cats' `===` does. So you can't do `x: SortedSet[A] === y: Set[A]`, but  `y: Set[A] === x: SortedSet[A]` will compile just fine. And given the symmetry law for `===` I wager this is fine.

File: tests/src/test/scala/cats/tests/NonEmptyChainSuite.scala [hunk 5]
```diff
@@ -60,118 +62,118 @@ class NonEmptyChainSuite extends NonEmptyCollectionSuite[Chain, NonEmptyChain, N
...
     forAll { (ci: NonEmptyChain[Int], i: Int) =>
-      ci.filterNot(_ === i).contains(i) should ===(false)
+      assert(ci.filterNot(_ === i).contains(i) === (false))
     }
   }
 
   test("Always nonempty after cons") {
     forAll { (ci: NonEmptyChain[Int], i: Int) =>
-      (i +: ci).nonEmpty should ===(true)
+      assert((i +: ci).nonEmpty === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptyChainSuite.scala:106]: alternatively also
```suggestion
      assert((i +: ci).nonEmpty)
```

File: tests/src/test/scala/cats/tests/NonEmptyChainSuite.scala [hunk 6]
```diff
@@ -60,118 +62,118 @@ class NonEmptyChainSuite extends NonEmptyCollectionSuite[Chain, NonEmptyChain, N
...
     forAll { (ci: NonEmptyChain[Int], f: Int => Boolean) =>
-      ci.filterNot(f).exists(f) should ===(false)
+      assert(ci.filterNot(f).exists(f) === (false))
     }
   }
 
   test("filter and then forall should always be true") {
     forAll { (ci: NonEmptyChain[Int], f: Int => Boolean) =>
-      ci.filter(f).forall(f) should ===(true)
+      assert(ci.filter(f).forall(f) === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptyChainSuite.scala:82]: an alternative could be
```suggestion
      assert(ci.filter(f).forall(f))
```

File: tests/src/test/scala/cats/tests/NonEmptyListSuite.scala [hunk 7]
```diff
@@ -84,86 +87,86 @@ class NonEmptyListSuite extends NonEmptyCollectionSuite[List, NonEmptyList, NonE
 
   test("Show is not empty and is formatted as expected") {
     forAll { (nel: NonEmptyList[Int]) =>
-      nel.show.nonEmpty should ===(true)
-      nel.show.startsWith("NonEmptyList(") should ===(true)
-      nel.show should ===(implicitly[Show[NonEmptyList[Int]]].show(nel))
-      nel.show.contains(nel.head.show) should ===(true)
+      assert(nel.show.nonEmpty === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptyListSuite.scala:90]: those asserting `=== true` could be
```suggestion
      assert(nel.show.nonEmpty)
```

File: tests/src/test/scala/cats/tests/NonEmptyListSuite.scala [hunk 8]
```diff
@@ -256,116 +259,116 @@ class NonEmptyListSuite extends NonEmptyCollectionSuite[List, NonEmptyList, NonE
...
     }
   }
   test("NonEmptyList#nonEmptyPartition remains sorted") {
     forAll { (nel: NonEmptyList[Int], f: Int => Either[String, String]) =>
       val sorted = nel.map(f).sorted
       val ior = Reducible[NonEmptyList].nonEmptyPartition(sorted)(identity)
 
-      ior.left.map(xs => xs.sorted should ===(xs))
-      ior.right.map(xs => xs.sorted should ===(xs))
+      assert(ior.left.map(xs => xs.sorted === (xs)).getOrElse(true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptyListSuite.scala:352]:
```suggestion
      assert(ior.left.forall(xs => xs.sorted === xs))
```

File: tests/src/test/scala/cats/tests/NonEmptyListSuite.scala [hunk 9]
```diff
@@ -256,116 +259,116 @@ class NonEmptyListSuite extends NonEmptyCollectionSuite[List, NonEmptyList, NonE
...
+      assert(ior.right.map(xs => xs.sorted === (xs)).getOrElse(true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptyListSuite.scala:353]:
```suggestion
      assert(ior.right.forall(xs => xs.sorted === xs))
```

File: tests/src/test/scala/cats/tests/NonEmptySetSuite.scala [hunk 10]
```diff
@@ -42,87 +44,86 @@ class NonEmptySetSuite extends CatsSuite {
...
     }
   }
 
   test("Show is not empty and is formatted as expected") {
     forAll { (nes: NonEmptySet[Int]) =>
-      nes.show.nonEmpty should ===(true)
-      nes.show.startsWith("NonEmptySortedSet(") should ===(true)
-      nes.show should ===(implicitly[Show[NonEmptySet[Int]]].show(nes))
-      nes.show.contains(nes.head.show) should ===(true)
+      assert(nes.show.nonEmpty === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptySetSuite.scala:55]:
```suggestion
      assert(nes.show.nonEmpty)
```

File: tests/src/test/scala/cats/tests/NonEmptySetSuite.scala [hunk 11]
```diff
@@ -42,87 +44,86 @@ class NonEmptySetSuite extends CatsSuite {
...
+      assert(nes.show.startsWith("NonEmptySortedSet(") === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptySetSuite.scala:56]:
```suggestion
      assert(nes.show.startsWith("NonEmptySortedSet("))
```

File: tests/src/test/scala/cats/tests/NonEmptySetSuite.scala [hunk 12]
```diff
@@ -42,87 +44,86 @@ class NonEmptySetSuite extends CatsSuite {
...
+      assert(nes.show === (implicitly[Show[NonEmptySet[Int]]].show(nes)))
+      assert(nes.show.contains(nes.head.show) === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/NonEmptySetSuite.scala:58]:
```suggestion
      assert(nes.show.contains(nes.head.show))
```

File: tests/src/test/scala/cats/tests/SortedMapSuite.scala [hunk 13]
```diff
@@ -44,9 +46,9 @@ class SortedMapSuite extends CatsSuite {
 
   test("show isn't empty and is formatted as expected") {
     forAll { (map: SortedMap[Int, String]) =>
-      map.show.nonEmpty should ===(true)
-      map.show.startsWith("SortedMap(") should ===(true)
-      map.show should ===(implicitly[Show[SortedMap[Int, String]]].show(map))
+      assert(map.show.nonEmpty === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/SortedMapSuite.scala:49]:
```suggestion
      assert(map.show.nonEmpty)
```

File: tests/src/test/scala/cats/tests/SortedMapSuite.scala [hunk 14]
```diff
@@ -44,9 +46,9 @@ class SortedMapSuite extends CatsSuite {
...
+      assert(map.show.startsWith("SortedMap(") === (true))
```
Comment (barambani) [inline tests/src/test/scala/cats/tests/SortedMapSuite.scala:50]:
```suggestion
      assert(map.show.startsWith("SortedMap("))
```

**Comment by LukaJCB** [comment]
Thanks for those, I think unblocking the dotty work is more important than fixing all the style issues for now, but I'm sure some regex master could fix this for us at some point 😄

**Comment by larsrh** [comment]
> I think unblocking the dotty work is more important than fixing all the style issues for now

For sure. Let's ship this :rocket:

**Comment by LukaJCB** [comment]
Alright, merging this :)
